### PR TITLE
compiler: Patch issue #1592 (HaloScheme with time subdimensions)

### LINKED ISSUE #1592: HaloScheme with time subdimensions
HaloScheme is breaking when a time subdimension is involved

```
from devito import *

grid = Grid(shape=(11, 11))

u = TimeFunction(name="u", grid=grid, time_order=2, space_order=4, save=10)

time_subsampled = ConditionalDimension('t_sub', parent=grid.time_dim, factor=2)

v = TimeFunction(name="v", grid=grid, space_order=4, time_dim=time_subsampled, save=5)
w = Function(name="w", grid=grid, space_order=4)

op = Operator(Eq(w, v.dx))
```

Leads to the error

```
    op = Operator(Eq(w, v.dx))
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/backends.py", line 73, in __call__
    return t(*args, **kwargs)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/operator/operator.py", line 157, in __new__
    op = cls._build(expressions, **kwargs)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/operator/operator.py", line 187, in _build
    stree = cls._lower_stree(clusters, **kwargs)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/tools/decorators.py", line 163, in __call__
    retval = self.func(*args, **kwargs)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/operator/operator.py", line 345, in _lower_stree
    stree = stree_build(clusters)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/ir/stree/algorithms.py", line 26, in stree_build
    stree = stree_make_halo(stree)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/ir/stree/algorithms.py", line 97, in stree_make_halo
    halo_schemes[n] = HaloScheme(n.exprs, n.ispace)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/mpi/halo_scheme.py", line 71, in __init__
    loc_indices = compute_local_indices(f, local, ispace, scope)
  File "/Users/mathiaslouboutin/research/London/CodeGen/devito/devito/mpi/halo_scheme.py", line 439, in compute_local_indices
    loc_indices[d] = func(candidates, key=lambda i: i-d)
  File "/usr/local/lib/python3.8/site-packages/sympy/core/relational.py", line 376, in __nonzero__
    raise TypeError("cannot determine truth value of Relational")
TypeError: cannot determine truth value of Relational
```

### PR DESCRIPTION
**Pull request by Leitevmd**
Fix #1592

### CODE REVIEW

**Comment by Leitevmd** [comment]
I think I made some mistake here. I'll review that and let you guys know.

**Comment by Leitevmd** [comment]
Some explanation..
When using the `factor` argument in `ConditionalDimension`, `LoweredEq` replaces the dimension index by `IntDiv(index, factor)`. Since `IntDiv` is a `sympy.Expr` child, builtins `max` and `min` are getting in trouble when selecting maximum/minimum `loc_indice` from candidates in `compute_local_indices(...)`. Replacing these builtins by `sympy.Max/Min` solved the issue (not sure if this is the only way). It just needed a little adjustment since `sympy.Max/Min` does not offer a key-like argument as in `max/min`.

**Comment by Leitevmd** [comment]
CI errors are not related with the issue, but with the way I'm checking the index in test_dimension.py. I'll take a look at this soon.

File: devito/mpi/halo_scheme.py [hunk 1]
```diff
@@ -428,13 +428,24 @@ def compute_local_indices(f, dims, ispace, scope):
...
         if d.is_Stepping:
-            loc_indices[d] = func(candidates, key=lambda i: i.origin - d)
+            loc_indices[d] = select(candidates, func=func, key=lambda i: i.origin - d)
         else:
-            loc_indices[d] = func(candidates, key=lambda i: i-d)
+            loc_indices[d] = select(candidates, func=func, key=lambda i: i - d)
     return loc_indices
+
+
+def select(candidates, func, key):
```
Comment (FabioLuporini) [inline devito/mpi/halo_scheme.py:443]: This smells like an overkill. Do we really need a whole new function to select from `candidates`? That sounds odd

Comment (Leitevmd) [reply devito/mpi/halo_scheme.py]: You're totally right!! Refactored using a dictionary.

File: tests/test_dimension.py [hunk 2]
```diff
@@ -597,6 +597,16 @@ def test_time_subsampling_fd(self):
                        usave[time_subsampled + 1, x, y],
                        usave[time_subsampled, x, y]]
 
+    def test_issue_1592(self):
```
Comment (mloubout) [inline tests/test_dimension.py:600]: Can we also have the case from the issue.

Comment (Leitevmd) [reply tests/test_dimension.py:600]: Sure! Just create the operator? Any execution/assertion?

Comment (FabioLuporini) [reply tests/test_dimension.py:600]: somehow assert the generated code is correct?

Comment (FabioLuporini) [reply tests/test_dimension.py:600]: ah, maybe you alreday are! I just see

```
         assert exprs[-1].expr.lhs.indices[0] == IntDiv(time, 2) + 1
```

**Review by FabioLuporini — COMMENTED** [review]
I think this LGTM. @mloubout ?

**Comment by mloubout** [comment]
GTG, merging

**Pull request merged**